### ISSUE TITLE
Boost 1.58

### ISSUE THREAD

**Issue by bgoodri** [stan-dev/stan#1457]
BH updated to 1.58, so we should probably fix the stanc issue that we have seen on the mailing list a few times when people are using their own more recent Boost and then update ourselves. The first error is
```
In file included from lang__grammars__expression07_grammar_inst.cpp:18:
In file included from /usr/local/lib/R/site-library/StanHeaders/include/stan/lang/grammars/expression07_grammar_inst.cpp:1:
In file included from /usr/local/lib/R/site-library/StanHeaders/include/stan/lang/grammars/expression07_grammar_def.hpp:16:
In file included from /usr/local/lib/R/site-library/BH/include/boost/spirit/include/qi.hpp:16:
In file included from /usr/local/lib/R/site-library/BH/include/boost/spirit/home/qi.hpp:14:
In file included from /usr/local/lib/R/site-library/BH/include/boost/spirit/home/qi/action.hpp:14:
In file included from /usr/local/lib/R/site-library/BH/include/boost/spirit/home/qi/action/action.hpp:14:
In file included from /usr/local/lib/R/site-library/BH/include/boost/spirit/home/qi/meta_compiler.hpp:14:
In file included from /usr/local/lib/R/site-library/BH/include/boost/spirit/home/support/meta_compiler.hpp:16:
In file included from /usr/local/lib/R/site-library/BH/include/boost/spirit/include/phoenix_limits.hpp:11:
In file included from /usr/local/lib/R/site-library/BH/include/boost/phoenix/core/limits.hpp:26:
In file included from /usr/local/lib/R/site-library/BH/include/boost/proto/proto_fwd.hpp:28:
/usr/local/lib/R/site-library/BH/include/boost/utility/result_of.hpp:189:46: error: too few template arguments for class template 'result'
struct result_of_nested_result : F::template result<FArgs>
```
after which the error log amounts to 2529 lines but the ones that specifically have "error:" in them are
```
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:41: error: no viable overloaded '='
/usr/local/lib/R/site-library/BH/include/boost/phoenix/core/detail/preprocessed/function_eval_10.hpp:171:57: error: no type named 'type' in 'boost::result_of<const stan::lang::addition_expr3 (stan::lang::expression &, stan::lang::expression &, std::basic_stringstream<char> &)>'
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:41: error: no viable overloaded '='
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:9: error: no type named 'result_type' in 'boost::proto::switch_<boost::phoenix::meta_grammar, boost::proto::tag_of<boost::proto::_> ()>::impl<const boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::phoenix::detail::tag::function_eval, boost::proto::argsns_::list4<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<stan::lang::addition_expr3>, 0>, boost::phoenix::actor<boost::spirit::attribute<0> >, boost::phoenix::actor<boost::spirit::argument<0> >, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::reference_wrapper<std::basic_stringstream<char> > >, 0> > >, 4> > &, boost::phoenix::vector4<const boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::assign, boost::proto::argsns_::list2<boost::proto::exprns_::expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::spirit::attribute<0> >, 0>, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::phoenix::detail::tag::function_eval, boost::proto::argsns_::list4<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<stan::lang::addition_expr3>, 0>, boost::phoenix::actor<boost::spirit::attribute<0> >, boost::phoenix::actor<boost::spirit::argument<0> >, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::reference_wrapper<std::basic_stringstream<char> > >, 0> > >, 4> > >, 2> > *, boost::fusion::vector1<stan::lang::expression &> &, boost::spirit::context<boost::fusion::cons<stan::lang::expression &, boost::fusion::cons<int, boost::fusion::nil_> >, boost::fusion::vector0<void> > &, bool &> &, const boost::phoenix::default_actions &>'
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:41: error: no viable overloaded '='
/usr/local/lib/R/site-library/BH/include/boost/phoenix/core/detail/preprocessed/actor_operator_10.hpp:71:20: error: no matching function for call to 'eval'
/usr/local/lib/R/site-library/BH/include/boost/utility/result_of.hpp:189:46: error: too few template arguments for class template 'result'
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:41: error: no viable overloaded '='
/usr/local/lib/R/site-library/BH/include/boost/phoenix/core/detail/preprocessed/function_eval_10.hpp:171:57: error: no type named 'type' in 'boost::result_of<const stan::lang::subtraction_expr3 (stan::lang::expression &, stan::lang::expression &, std::basic_stringstream<char> &)>'
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:41: error: no viable overloaded '='
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:9: error: no type named 'result_type' in 'boost::proto::switch_<boost::phoenix::meta_grammar, boost::proto::tag_of<boost::proto::_> ()>::impl<const boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::phoenix::detail::tag::function_eval, boost::proto::argsns_::list4<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<stan::lang::subtraction_expr3>, 0>, boost::phoenix::actor<boost::spirit::attribute<0> >, boost::phoenix::actor<boost::spirit::argument<0> >, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::reference_wrapper<std::basic_stringstream<char> > >, 0> > >, 4> > &, boost::phoenix::vector4<const boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::assign, boost::proto::argsns_::list2<boost::proto::exprns_::expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::spirit::attribute<0> >, 0>, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::phoenix::detail::tag::function_eval, boost::proto::argsns_::list4<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<stan::lang::subtraction_expr3>, 0>, boost::phoenix::actor<boost::spirit::attribute<0> >, boost::phoenix::actor<boost::spirit::argument<0> >, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::reference_wrapper<std::basic_stringstream<char> > >, 0> > >, 4> > >, 2> > *, boost::fusion::vector1<stan::lang::expression &> &, boost::spirit::context<boost::fusion::cons<stan::lang::expression &, boost::fusion::cons<int, boost::fusion::nil_> >, boost::fusion::vector0<void> > &, bool &> &, const boost::phoenix::default_actions &>'
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:41: error: no viable overloaded '='
/usr/local/lib/R/site-library/BH/include/boost/phoenix/core/detail/preprocessed/actor_operator_10.hpp:71:20: error: no matching function for call to 'eval'
/usr/local/lib/R/site-library/BH/include/boost/utility/result_of.hpp:189:46: error: too few template arguments for class template 'result'
/usr/local/lib/R/site-library/BH/include/boost/phoenix/core/detail/preprocessed/function_eval_10.hpp:121:52: error: no type named 'type' in 'boost::result_of<const stan::lang::validate_expr_type3 (stan::lang::expression &, std::basic_stringstream<char> &)>'
/usr/local/lib/R/site-library/BH/include/boost/proto/transform/default.hpp:154:9: error: no type named 'result_type' in 'boost::proto::switch_<boost::phoenix::meta_grammar, boost::proto::tag_of<boost::proto::_> ()>::impl<const boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::phoenix::detail::tag::function_eval, boost::proto::argsns_::list3<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<stan::lang::validate_expr_type3>, 0>, boost::phoenix::actor<boost::spirit::attribute<0> >, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::reference_wrapper<std::basic_stringstream<char> > >, 0> > >, 3> > &, boost::phoenix::vector4<const boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::assign, boost::proto::argsns_::list2<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::phoenix::argument<3> >, 0>, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::phoenix::detail::tag::function_eval, boost::proto::argsns_::list3<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<stan::lang::validate_expr_type3>, 0>, boost::phoenix::actor<boost::spirit::attribute<0> >, boost::phoenix::actor<boost::proto::exprns_::basic_expr<boost::proto::tagns_::tag::terminal, boost::proto::argsns_::term<boost::reference_wrapper<std::basic_stringstream<char> > >, 0> > >, 3> > >, 2> > *, boost::fusion::vector1<boost::spirit::unused_type &> &, boost::spirit::context<boost::fusion::cons<stan::lang::expression &, boost::fusion::cons<int, boost::fusion::nil_> >, boost::fusion::vector0<void> > &, bool &> &, const boost::phoenix::default_actions &>'
/usr/local/lib/R/site-library/BH/include/boost/phoenix/core/detail/preprocessed/actor_operator_10.hpp:71:20: error: no matching function for call to 'eval'
```

**Comment by syclik** [comment]
Should Stan bump its version to match BH? Maybe it'll be easier to figure
out if a few of us are looking at it.

**Comment by bgoodri** [comment]
We could make a branch for it. I think the build problems are just with
stanc.

**Comment by bob-carpenter** [comment]
OK, I'll see what I can find.

**Comment by bob-carpenter** [comment]
I have this one passing the parse unit tests now on my machine.

I'll run the rest of the tests and put in a pull request tomorrow 
and hope all goes well on Windows.  

I found the patch I needed in this diff:

  https://launchpadlibrarian.net/184598940/hugin-2013-boost-156.diff

from this bug tracker: 

  https://bugs.launchpad.net/hugin/+bug/1368478

**Comment by bob-carpenter** [comment]
Awesome!  Thanks for finishing the fixes for Boost 1.58.  

I think the next version should be Stan 2.7.0 because there's
all the math lib changes as well as the new Boost lib.  

- Bob

**Comment by syclik** [comment]
You're welcome. I just updated CmdStan to match.

Math lib: I'll try to get as much done tomorrow so we can release it asap.
